Feat/172 helper message component

manvinderjit opened the pull request:
## Summary

Need helpers for text inputs, selects or other interactive elements, to provide additional information and, optionally, status regarding the placed content. Helpers can give input-related hints and also indicate whether the user has successfully filled the input, or if there's still any mistake yet to fix.

## Tests

Pending

## Additional information

Closes #172 

Adds helpers based on the following figma design:

<img width="463" height="619" alt="image" src="https://github.com/user-attachments/assets/b2500442-8ffb-4463-af82-a47547198532" />

manvinderjit committed:
feat: add HTML and types for HelperMessage element

add HTML structure and types for HelperMessage component.

manvinderjit committed:
feat: add css styles for HelperMessage element

bcha92 commented on src/components/HelperMessageComponent/HelperMessageComponent.tsx:
```diff
@@ -0,0 +1,19 @@
+const HelperMessageComponent = ({ variant = 'default', labelText }: Props) => {
+	return (
+		<div className='helper-container' data-variant={variant}>
+			<span className='helper-icon' />
```
Not a blocker per se, but given we have `svg` icons to load for the `.helper-icon` elements and for the most part they don't necessarily need to be loaded at the start of the page, would implementing lazy loading be useful or beneficial in terms of speed and FCP (First Contentful Paint) in a reusable component like this one?
```jsx
<span className="helper-icon" loading="lazy" />
```

But otherwise I think it looks good to go 👍

madcampos replied:
We are using react, FCP is already an issue on itself 🙃

That being said, I don't think it will particularly help as there is no easy way to add lazy loading to SVG elements, there is a native attribute for both images and iframes, but not for SVGs.

If we would implement it ourselves, there would be added complexity in the form of creating an intersection observer, then removing it when the svg was loaded, then adding some request for the svg content, all while keeping track of things using react and state.

One very viable option instead would be to have a shared svg definition added to the page, and then each icon could be only a `use` tag. That removes duplication, while still allowing for flexibility.

manvinderjit replied:
@madcampos You mean like this 
`<svg className="icon">
    <use href="" />
  </svg>`

madcampos replied:
Yes. It is a combo:

First the defs as a global element.
```html
<html>
  <body>
    <svg id="icon-definitions" aria-hidden="true" role="none">
      <defs>
        <!-- All icon defs go here -->
      </defs>
    </svg>
    ...
  </body>
</html>
```

Then the component will use the definitions as you said:
```html
<svg class="icon" data-icon="..."><use href="#..." /></svg>
```

The caveats to that are:
- The first SVG still takes space in page, so we need to hide it with CSS.
- The first SVG cannot be completely hidden with things like `display: none` or else effects, filters, masks, and other things will not work, so it needs to be hidden using a technique similar to the sr only class.
- It doesn't work inside web components as the id will not exist inside the Shadow DOM.

manvinderjit replied:
@madcampos Considering we have a multi page app, should we build a separate component for svg definitions and reuse it on each page?

madcampos replied:
That works, yes.

The ideal solution would be to statically include the svg, but for now that works.

manvinderjit replied:
Added the definitions as a Component.

kbventures commented on src/components/HelperMessageComponent/HelperMessageComponent.css:
```diff
@@ -0,0 +1,59 @@
+@layer components {
```
Nitpick:  Add a simple one to a page so we can see what it looks like.  Great work btw.

manvinderjit replied:
@kbventures Do you want me to add a separate page for these or include them in the inputs page?

manvinderjit replied:
Done

madcampos commented on src/components/HelperMessageComponent/HelperMessageComponent.css:
**nitpick/question**: Why don't we use component icons instead of the background images?

With live SVGs inside the DOM we can target the icons with CSS and change things like color and animate them.

manvinderjit replied:
@madcampos We used to add component icons but you recommended to use them as background images. Where do you recommend we use the former vs the latter?

madcampos replied:
Background images work for things like the arrow on a select, or as the list indicator for a list item. Those are situations where the image is an enhancement to a single element.

For components with more than one element, then using an SVG is usually better.

The other point is styling, if we need to control or change how the icon will be styled (which is almost always), then using SVG is better as we can style it with CSS. An SVG image cannot have dynamic styles.

manvinderjit committed:
feat: add helper message svgs definition component

Add a component that has svg definitions for the helperMessages

manvinderjit committed:
style: add css for helper svg definitions

manvinderjit committed:
fix: update helper message to use svg defs

manvinderjit committed:
fix: update helper message css

manvinderjit committed:
feat: add usage samples for helper message

manvinderjit committed:
fix: fix confirm account path in vite config

madcampos commented on src/components/HelperMessageComponent/HelperMessageSvgDefinitions.tsx:
```diff
@@ -0,0 +1,74 @@
+							strokeLinecap='round'
+							strokeLinejoin='round'
+						/>
+						<path d='M8 10.6667V8' stroke='#3388FF' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+						<path d='M8 5.3335H8.00667' stroke='#3388FF' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+					</g>
+					<defs>
+						<clipPath id='clip-info'>
+							<rect width='16' height='16' fill='white' />
+						</clipPath>
+					</defs>
+				</symbol>
+
+				{/* Error (X Icon) */}
+				<symbol id='icon-error' viewBox='0 0 16 16' fill='none'>
+					<path d='M12 4L4 12' stroke='#E51429' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+					<path d='M4 4L12 12' stroke='#E51429' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+				</symbol>
+
+				{/* Success (Checkmark) */}
+				<symbol id='icon-success' viewBox='0 0 16 16' fill='none'>
+					<path d='M13.3334 4L6.00002 11.3333L2.66669 8' stroke='#06D6A0' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+				</symbol>
+
+				{/* Warning */}
+				<symbol id='icon-warning' viewBox='0 0 16 16' fill='none'>
+					<path
+						d='M14.4867 12L9.15335 2.66665C9.03706 2.46146 8.86842 2.29078 8.66463 2.17203C8.46084 2.05329 8.22921 1.99072 7.99335 1.99072C7.75749 1.99072 7.52585 2.05329 7.32206 2.17203C7.11828 2.29078 6.94964 2.46146 6.83335 2.66665L1.50001 12C1.38247 12.2036 1.32083 12.4346 1.32135 12.6697C1.32187 12.9047 1.38453 13.1355 1.50298 13.3385C1.62142 13.5416 1.79145 13.7097 1.99581 13.8259C2.20018 13.942  ...
+						stroke='#DE9502'
+						strokeWidth='2'
+						strokeLinecap='round'
+						strokeLinejoin='round'
+					/>
+					<path d='M8 6V8.66667' stroke='#DE9502' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+					<path d='M8 11.3335H8.00667' stroke='#DE9502' strokeWidth='2' strokeLinecap='round' strokeLinejoin='round' />
+				</symbol>
+			</defs>
+		</svg>
+	);
+};
```
**suggestion**: we can make it a simple script, react doesn't have to _know_ about this so it won't have it's feelings hurt.

```typescript
export function initializeIconDefinitions() {
  if (!document.querySelector('#icon-definitions')) {
    document.body.insertAdjacentHtml('afterbegin',
      // SVG string goes here
    );
  }
}
```
 
Then on the component we `import` this function and call it:
```typescript
import { initializeIconDefinitions } from './icon-definitions.ts';

export function RandomIconComponent() {
  initializeIconDefinitions();

  //...
}
```

If you want to even go further, there is another option to use the module import as a side effect and take advantage of the top level `await` in modules.

The definition module would be rewritten as such:
```typescript
// No export here as it will be called and used as a side effect.
// Note the function is also now async.
async function initializeIconDefinitions() {
  if (!document.querySelector('#icon-definitions')) {
    const result = await fetch('/icon-defs.svg');
    const svgText = await result.text();

    document.body.insertAdjacentHtml('afterbegin', svgText);
  }
}

// Top level await in modules is supported in browsers for a while.
await initializeIconDefinitions();
```

Then the import inside the component would just be:
```typescript
// This import definition is executed for side effects.
// Modules are async due to network fetching and all, so it Just Works™
import './icon-definitions.ts';

// ...
```

manvinderjit replied:
@madcampos Updated the code using the latter.

madcampos commented on src/pages/usage-helper-message/sample-helper-message.tsx:
```diff
@@ -0,0 +1,80 @@
+			<div className='App'>
+				{/* NOTE: This is temp UI to test the helper message usage with text input or other components */}
+
+				<h2 style={{ textAlign: 'center' }}>Render the Design System's Helper Message Input Component Cases</h2>
```
**issue**: there is no `h1` on this page.

manvinderjit replied:
I generally leave the h1 for the index.html page instead of the component, but I can change it because we are not going to use it here.

madcampos commented on src/pages/usage-helper-message/sample-helper-message.tsx:
```diff
@@ -0,0 +1,80 @@
+				</p>
+
+				<div className='component-container'>
+					<h3>
+						Default Variant:
+					</h3>
+					<HelperMessageComponent variant='default' labelText='Helper Message' />
```
```suggestion
					<fieldset>
						<legend>Default Variant</legend>
						<HelperMessageComponent variant='default' labelText='Helper Message' />
					</fieldset>
```

For those scenarios where you are exposing sets of components and there is a need for grouping it is better to use the [`<fieldset>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/fieldset) or the [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/details) (with the `open` attribute set) elements.

The `<details>` element is preferred for longer lists or where the section contains a lot of information, so it can be collapsed. But for just laying out multiple states and scenarios, prefer the `<fieldset>` element

madcampos commented on src/pages/usage-helper-message/style.css:
```diff
@@ -0,0 +1,28 @@
+		font-weight: var(--weight-bold);
+	}
+
+	.component-container {
```
**issue**: please use a more descriptive name to avoid clashes and improve clarity.

manvinderjit committed:
fix: update html and css for semantics

replace h3 html tags with fieldset and legend.
update css class name.
remove non-required css.
update heading hierarchy.

manvinderjit committed:
refactor: use js script for helper msg svg defs
